api/src/test/java/org/openmrs/api/impl/DiagnosisServiceImplTest.java

```diff
@@ -330,5 +345,228 @@ public void purgeDiagnosis_shouldPurgeDiagnosis() {
+				.getDiagnosisAttributeTypeByUuid("949daf5b-a83e-4b65-b914-502a553243d3");
+		DiagnosisAttribute diagnosisAttribute = new DiagnosisAttribute();
+		diagnosisAttribute.setAttributeType(diagnosisAttributeType);
+		diagnosisAttribute.setCreator(Context.getUserService().getUser(1));
+		diagnosisAttribute.setVoided(false);
+		diagnosisAttribute.setValueReferenceInternal("Diagnosis Attribute Reference");
+		final String UUID = "3c6422d8-7bdd-4e20-bd1b-a590f084db08";
+		Condition condition = conditionService.getConditionByUuid("2cc6880e-2c46-15e4-9038-a6c5e4d22fb7");
+		Encounter encounter = encounterService.getEncounterByUuid("y403fafb-e5e4-42d0-9d11-4f52e89d123r");
+		Patient patient = patientService.getPatient(2);
+		Diagnosis diagnosis = new Diagnosis();
+		diagnosis.setUuid(UUID);
+		diagnosis.setEncounter(encounter);
+		diagnosis.setCondition(condition);
+		diagnosis.setCertainty(ConditionVerificationStatus.CONFIRMED);
+		diagnosis.setPatient(patient);
+		diagnosis.setRank(1);
+		diagnosis.setVoided(false);
+		diagnosis.setFormField(NAMESPACE, FORMFIELD_PATH);
+		diagnosis.addAttribute(diagnosisAttribute);
+		diagnosisService.save(diagnosis);
+		assertNotNull(diagnosis.getId(), "Successfully Saved Diagnosis");
+		Diagnosis savedDiagnosis = diagnosisService.getDiagnosisByUuid(UUID);
+		assertEquals(condition, savedDiagnosis.getCondition());
+		assertEquals(encounter, savedDiagnosis.getEncounter());
+		assertEquals(patient, savedDiagnosis.getPatient());
+		assertEquals(ConditionVerificationStatus.CONFIRMED, savedDiagnosis.getCertainty());
+		assertEquals(1, savedDiagnosis.getRank());
+		assertEquals(false, savedDiagnosis.getVoided());
+		assertEquals(NAMESPACE + "^" + FORMFIELD_PATH, savedDiagnosis.getFormNamespaceAndPath());
+		assertThat(savedDiagnosis.getAttributes(), hasItem(diagnosisAttribute));
+		for (DiagnosisAttribute attribute : savedDiagnosis.getAttributes()) {
+			if (attribute.getAttributeType().equals(diagnosisAttributeType)) {
+				assertThat(attribute.getAttributeType().getName(), is("Differential Diagnosis"));
+			}
+		}
+	}
+
+	/**
+	 * @see org.openmrs.api.DiagnosisService#save(Diagnosis)
+	 */
+	@Test
+	public void saveDiagnosis_shouldEditTheExistingDiagnosisRemovingTheAssociatedAttributesWhenRequired() {
+		final int RANK = 5;
+		final DiagnosisAttribute diagnosisAttribute = Context.getDiagnosisService()
+				.getDiagnosisAttributeByUuid("31f7c3cd-699b-4ed3-af10-563e024cae76");
+		final DiagnosisAttributeType diagnosisAttributeType = Context.getDiagnosisService()
+				.getDiagnosisAttributeTypeByUuid("949daf5b-a83e-4b65-b914-502a553243d3");
+		Diagnosis diagnosis = diagnosisService.getDiagnosis(1);
+		assertEquals(ConditionVerificationStatus.CONFIRMED, diagnosis.getCertainty());
+		assertEquals(1, diagnosis.getRank());
+		assertEquals(false, diagnosis.getVoided());
+		assertThat(diagnosis.getAttributes(),
+				both(hasItem(diagnosisAttribute)).and(contains(allOf(hasProperty("valueReference", is("Testing Reference"))))));
```

Comment:
Wouldn't it be enough to remove the attribute and then just assert that it no longer exists after saving and fetching the diagnosis again?

Reply:
Fixed